### DISCUSSION TITLE
Add a ruler to Xournalpp

### DISCUSSION THREAD

**Discussion by MrPancakes39** [xournalpp/xournalpp#4254] [Ideas] [answered]
Hello,
So Microsoft have a Ruler in Snip and Sketch that a lot of my engineering friends find it useful, it would be nice to have that in Xournalpp too!
So I will try in my spare time to write a plugin to do add it, not sure if that is possible but will try!
Image of Microsoft Ruler:
![Image of Microsoft Ruler](https://user-images.githubusercontent.com/41960961/192159967-5c354e6e-92e9-4395-920e-461cbd508e67.png)

**Comment by rolandlo** [comment, marked as answer]
Have you noticed that Xournal++ has a set square with more functionality than OneNote's ruler? Try it out (via the [nightly release](https://github.com/xournalpp/xournalpp/releases/tag/nightly)). Does that satisfy your needs?

![grafik](https://user-images.githubusercontent.com/40485433/192160456-e1641c83-0699-4e35-9551-7c00b387e402.png)

**Reply by MrPancakes39** [reply]
Looks really good and all that's needed. I'll try it asap, and tell you my feedback thanks fo much for bringing this to my attention. ❤️

**Reply by MrPancakes39** [reply]
@rolandlo I tried it and it's prefect! (in concept) I don't wanna seem like I'm complaining since you have obviously put a lot of work into this and I appreciate it! But I do have some criticism:
1) You should be able to rotate the setsquare. (maybe there is a way but I couldn't figure it out.)
2) Angle locking is great but I suggest to have it under a shortcut like while holding Alt or Shift key.

**Reply by rolandlo** [reply]
Of course you can rotate the set square, either via touch (2 fingers) or via keyboard (r/R). See https://github.com/xournalpp/xournalpp.github.io/blob/58ba5c989108f3e81019164391c595f306c250ac/docs/guide/tools/geometry.md
What do you mean with angle locking? Maybe angle snapping near straight lines?
You can prevent angle snapping near straight lines if you move the setsquare with 2 fingers (one on the set square, the other outside).

**Reply by MrPancakes39** [reply]
1) Sorry for my stupidity should have guessed r/R for rotation.
2) Yes, angle snapping I don't have a touch device only a mouse and yes if I have the endpoint is outside the setsquare no snapping occurs, but inside it snapping occurs almost randomly (or that's how it feels) that's why I suggested a shortcut to force snapping on current angle when pressed.

**Reply by rolandlo** [reply]
Good to know that you don't have a touch device.
1. Note by the way, that you can use `Alt` to get smaller increments. 
2. I'm still not sure what you mean. Angle snapping is only related to touch (when you move the set square near an existing straight line, it gets rotated such that it is perfectly aligned to the line, in case the set square is almost aligned to it already). Maybe you mean drawing (with the mouse or the pen), when you get straight lines aligned to the longest side of the setsquare in case you start inside the set square and at most 0.5cm away from the longest side of the set square. It's not random at all, it depends on where you start, if it is inside the set square and how close it is to the longest side of the set square (0.5 cm max).

**Reply by MrPancakes39** [reply]
1) This is the most useful and coolest setsquare I have ever seen, I can't believe you do this for free you guys are amazing!
2) > use `Alt` to get smaller increments.

That's excellent, I would recommend saying in the docs that : "You may need to disable Alt+R shortcut used by NVIDIA GeForce Experience"

3) > such that it is perfectly aligned to the line

You guys thought about everything!

4) Maybe I am not understanding how snapping works? I turned off grid and rotation snapping and recorded a video showing what I am talking about. [Video Here.](https://drive.google.com/file/d/1zvF0pOxLTAwVeKP74HNmiwJSz5UQl7Dm/view?usp=sharing)

Note: Although I am showing Xournal on Windows I use Xournal both on Windows and Linux so if you want me to test it on Linux I can do that too

**Reply by rolandlo** [reply]
That's what I mean:

[SetsquareSnapping.webm](https://user-images.githubusercontent.com/40485433/192289677-2e248d32-c5f5-4c21-bd00-9878eff9ca02.webm)

It helps to align the set square exactly to a straight line that is already drawn, useful in geometric constructions. You only get it when moving the set square via the touchscreen (which you don't have).

What your video shows is drawing lines using the set square (aligned to the longest side of the set square or with one end point locked to the mid point of the longest side of the set square). You only get that when you start drawing/writing inside the set square and not further than 0.5 cm from one of its sides. If you start drawing/writing somewhere else (until you release the pen), you will draw/write as normal (without set square)

**Reply by MrPancakes39** [reply]
I do understand that and I don't want to be wasting your time. I am talking what happens 0:20 and forwards where the line seem to "lock at a specific angle". The "locking" feels random to me, but with you having seen the video and not noticed anything weird about the mechanic I conclude that it's normal.

Edit: Maybe I can elaborate more for example If I want it to "lock" at say 26deg I can't do that, but it may look at like 20 deg then I can't "unlock" it. That's why I said it feels random to me. But it's probably me not understanding something.

**Reply by rolandlo** [reply]
You are not wasting my time. This is an implementation detail I already forgot about and is missing in the documentation. The locking happens when you move the cursor inside the (violet) semicircle with the angular marks. You can't see what precise angle you have when the whole line is inside the semicircle, so that's why it is locked once you get across the semicircle, so you can draw it precisely.

**Reply by MrPancakes39** [reply]
Oh I'm glad I'm not wasting your time, one last thing when it's locked and you try to go behind the midpoint it doesn't follow the mouse, it looks like some absolute value is taking effect, is that intentional? And about locking in the air like what happens at 0;34 is that also intentional?

Edit: And oh thanks I know understand how locking works <3

**Reply by rolandlo** [reply]
Those two behaviors are not intentional. I didn't really care about these cases so far. Thanks for reporting.

**Reply by MrPancakes39** [reply]
Glad I was able to help! I also noticed 1 more thing, when I switched my keyboard language from English to Arabic, the keybindings seem to have stopped working not all keybindings `Ctrl+D` and `Ctrl+Shift+P` these seem to work, only setsquare keybindings seem to stop `s/S` and `r/R` (with and without `Alt`) so that's also a bug.

**Reply by MrPancakes39** [reply]
Oh and @rolandlo thanks so much for your help and this amazing setsquare it's truly stunning I appreciate your work so much and wish to be a great developer like you in the future! Enjoy your day!

**Comment by dreng** [comment]
@rolandlo I just tried the set square feature. Amazing work, thank you very much! There are some things that should be improved though. Here's a list of what I found out after playing around on a Thinkpad X13 Yoga with touchscreen (I only tested this way - 2 fingers and a pen) :

- I could not get the snapping feature to work you showed in you video above. The set square never snapped in, no matter what. Is there anything I have to set in the preferences?
- There should be an option to snap the angle of the set square by steps of 1, 5 or 15 degrees or something like that. Otherwise, it is sometimes difficult to set an exact degree.
- While drawing a line along the set square, the position and angle of the set square is locked. That's absolutely logical and necessary. After that I had a hard time releasing the set square. I always tapped a few times or tapped on the toolbar and after that I was able to move the set square again. I've never found out what the final action was that released the set square. This approach was very tedious.
- Seems to be a bug: I added a new page and moved the set square down to that new page. When I zoomed the page the set square started to flicker at approx. 220% and then suddenly disappeared. It reappeared after zooming out again. In some cases, it was no longer possible to draw new lines.
- You mentioned moving the set square with 2 fingers while putting one finger outside of the set square. I you do this, the page is moving as well which is hard to use and seems to be not intended.

**Reply by rolandlo** [reply]
@dreng Thanks for the feedback and praise. :smile: I will add some answers.

> @rolandlo I just tried the set square feature. Amazing work, thank you very much! There are some things that should be improved though. Here's a list of what I found out after playing around on a Thinkpad X13 Yoga with touchscreen (I only tested this way - 2 fingers and a pen) :
> 
>     * I could not get the snapping feature to work you showed in you video above. The set square never snapped in, no matter what. Is there anything I have to set in the preferences?
>
No, nothing to set in the preferences. Maybe there is some issue with absolute and relative coordinates, which other users have experienced in different situation (pinch gesture for zooming, as I recall)
 
>     * There should be an option to snap the angle of the set square by steps of 1, 5 or 15 degrees or something like that. Otherwise, it is sometimes difficult to set an exact degree.

You mean while using touch? 

> 
>     * While drawing a line along the set square, the position and angle of the set square is locked. That's absolutely logical and necessary. After that I had a hard time releasing the set square. I always tapped a few times or tapped on the toolbar and after that I was able to move the set square again. I've never found out what the final action was that released the set square. This approach was very tedious.
> 
Hmm, I haven't experiences that. Does releasing the fingers from the screen and putting them on the screen again not suffice?

>     * Seems to be a bug: I added a new page and moved the set square down to that new page. When I zoomed the page the set square started to flicker at approx. 220% and then suddenly disappeared. It reappeared after zooming out again. In some cases, it was no longer possible to draw new lines.
> 
The set square is currently only supposed to be used on the page where it was activated. If you need it on a different page, just deactivate it and activate it on the new page.

>     * You mentioned moving the set square with 2 fingers while putting one finger outside of the set square. I you do this, the page is moving as well which is hard to use and seems to be not intended.

Again in my case it works fine, the page is not moving. Maybe also a problem with absolute and relative coordinates. It's a bit hard to debug, since this problem doesn't appear for me. I will have to read some old issues connected to that. 

It would help if you could add some short videos which highlight the problems you experienced with comments what you did and what is wrong.

**Reply by dreng** [reply]
> It would help if you could add some short videos which highlight the problems you experienced with comments what you did and what is wrong.

Sure.

> No, nothing to set in the preferences. Maybe there is some issue with absolute and relative coordinates, which other users have experienced in different situation (pinch gesture for zooming, as I recall)

I don't see any connection there. But who knows?
EDIT: Just found out that snapping does not work when using two fingers. If I lift one of the fingers the set square snaps as intended.

https://user-images.githubusercontent.com/20901110/193910304-877281f9-ad36-47ce-aaa1-0391797b1250.mp4

> You mean while using touch?

Exactly. It is very hard to turn the set square to a certain angle. Here's a video showing the problem:

https://user-images.githubusercontent.com/20901110/193909219-71b93c70-4ee2-4e09-93d4-626ad7c0a626.mp4

> Hmm, I haven't experiences that. Does releasing the fingers from the screen and putting them on the screen again not suffice?

No, it doesn't. But it turned out, that this is a side effect of the palm detection that was enabled in the preferences. After disabling palm detection everything worked fine. I provide a video anyway:

https://user-images.githubusercontent.com/20901110/193911695-a60022ec-0c75-4c83-bcae-4d85f34a7159.mp4

> Again in my case it works fine, the page is not moving. Maybe also a problem with absolute and relative coordinates. It's a bit hard to debug, since this problem doesn't appear for me. I will have to read some old issues connected to that.

Let me know if (and how) I can help. Here's what it looks like:

https://user-images.githubusercontent.com/20901110/193913292-2573c0a4-ffff-47dd-a94e-2a867e9c2a0a.mp4

**Reply by rolandlo** [reply]
Thanks a lot for the helpful videos. Some comments:
- `linesnap.mp4`: You should use one finger touch to get the snapping. It does not work with two fingers.
- `snapping.mp4`: I agree, this is hard with touch and deserves to be improved in future. It's easy with the keyboard (using R/r keys and the Alt modifier), but that does not help you, if you would like to use only touch/pen and have no buttons/wheels that you could map to these keys
- `palmdetection.mp4`: I haven't tested with palm detection as of now. Thanks for finding out about it. That can be improved.
- `scrollisue.mp4`: I can reproduce the issue now. Thanks for reporting. I will try to understand why that happens and what can be done about it.

**Reply by dreng** [reply]
>     * `scrollisue.mp4`: I can reproduce the issue now. Thanks for reporting. I will try to understand why that happens and what can be done about it.

I just found out the exact reason. It depends which of the two fingers touches first. This is only barely noticeable because both fingers are seemingly touching the screen at the same time.
If you touch the set square with the first finger first and the paper with the second finger after that, everything will be fine. But if you touch the paper first, it comes to the described issue.

**Comment by rolandlo** [comment]
I have pushed some fixes to the Setsquare refactor branch #4211. These two bugs should be fixed:

1) 2-finger scrolling (with first finger touching outside of the set square)
2) radius angle locking (when moving the cursor outside of the set square)

I couldn't reproduce the palm recognition bug related to the set square (touch disabling turned on). If you want to try and give feedback, try it out via the links from [this post](https://github.com/xournalpp/xournalpp/pull/4211#issuecomment-1302281279)

**Reply by dreng** [reply]
Thank you for working on this!

1. Seems to be resolved. It still depends on whether you put the first finger on the sheet or the set square first. However, the sheet now scrolls exactly with the set square and not at different speeds.
2. I don't know if I'm understanding you correctly, but I can't see anything new there. I am also not sure what is meant by "cursor". From my point of view, that word doesn't make any sense in touch mode. Maybe we were talking about different things. My problem was that I found it extremely difficult to set a certain angle. In the video, I wanted to set 15 degrees, but I never managed to hit this value exactly, but always overshot the target value at the slightest movement. It might help if there were some kind of grid so that only whole numbers of degrees or 5-degree increments could be set. Normally, holding the Shift or Ctrl buttons would be suitable for this, but not in plain touch mode.

Generally speaking, something is still missing in the haptics. If you compare how quickly you can work with the set square on real paper, there is still some room for improvement. There is a general lack of fine motor skills and therefore it always takes a long time to position the set square exactly. It gets really bad when you try to align the set square 90 degrees to a line and move the zero point exactly to the end of the line. This is much faster and more accurate on real paper.
I don't want to list all possible things that feel better on real paper. Just compare yourself and you'll get an idea. All I want to say is that a few more intelligent aids are needed to turn a good tool into a brilliant tool.

**Reply by dreng** [reply]
Oh, and just one more thing: you can hardly recognise the icon and it doesn't fit in well with the rest of the icons. But that is only a small, cosmetic thing.

**Reply by rolandlo** [reply]
1. I am aware, that it still depends on which finger you put on the page (outside the set square) or set square first. This makes sense to me though, since a (first) finger on the page will move the page and that shouldn't be stopped when putting a second finger on the set square in my opinion.
2. We are talking about different things.  I was talking about [this comment](https://github.com/xournalpp/xournalpp/discussions/4254#discussioncomment-3734631) by @MrPancakes39, which can be seen in the video. You are right that nothing is changed about locking the set square to a precise angle. 

I also agree that something is still missing in the haptics and that in most situations it's clearly easier to position a real set square on real paper. This can be improved, for sure, but we will have to figure out how exactly. I don't agree with the precise situation you mentioned though. The digital setsquare is auto-rotated to be exactly perpendicular to a line on the current layer when the zero point comes close to the line (as long as the absolute difference in angles is at most 3 degrees). You can't do that as precisely on real paper. Moving the digital set square with one finger such that the zero point (more or less) exactly matches the end of the line is not time-consuming either. 

Yes, the icons are bad, and I hope someone with much better designing skills (like @Fatih20 ) will eventually improve it. Frankly speaking, I don't find any pleasure in designing icons, although I love to have beautiful icons, too.

Note that the set square (and even more so the compass) are still in an early developing stage. It will improve over time, with the feedback and ideas of the community.

**Reply by dreng** [reply]
1. Agree
2. Got it

Just to make that clear: Please don't get me wrong. Everything I comment here should be understood as feedback, suggestions and personal opinions. I am very aware of the fact that the branch is in an early developing stage and I am very thankful that you are working on that feature. I don't intend to push anything.

Regarding the icons, you might want to use these:
Edit: I removed the icons here after a small overhaul and a pull request. See https://github.com/xournalpp/xournalpp/pull/4422

Not a DaVinci piece either, but at least visible ;-) and it fits better into the scheme. You may use them under GNU license or whatever license you use for all the other icons. I can either make a pull request if desired.

**Reply by rolandlo** [reply]
Yes, they look clearly better. Maybe the colored icons need more work with the actual coloring, but the Lucide icons (both dark and light) look good to me.
Please make a pull request and add the license info to the [`copyright.txt`](https://github.com/xournalpp/xournalpp/blob/master/copyright.txt) file as part of the pull request. Note that the pull request will only get merged after the compass made it into the master branch, which may take a while.

**Comment by tattsan** [comment]
This delay is also seen when I use touch scrolling. So this is not an issue of the setsquare, but of the hand recogintion.
Even if the stylus is away from the screen, scrolling by touch will not start if the pen icon remains where it was released.

**Reply by rolandlo** [reply]
I have started investigating this and made some first conclusions (`HandRecognition::event` only cares about the device, not about the event type; device blocking sometimes seems to take longer than expected; Gdk sometimes does not emit a `TOUCH_BEGIN` event when it recognizes a swipe gesture). This will have to be discussed in a new issue/PR though.
